stdlib/LinearAlgebra/src/hessenberg.jl

```diff
@@ -94,17 +94,96 @@ Base.copy(A::Transpose{<:Any,<:UpperHessenberg}) = tril!(transpose!(similar(A.pa
 rmul!(H::UpperHessenberg, x::Number) = (rmul!(H.data, x); H)
 lmul!(x::Number, H::UpperHessenberg) = (lmul!(x, H.data); H)
 
-# (future: we could also have specialized routines for UpperHessenberg * UpperTriangular)
-
 fillstored!(H::UpperHessenberg, x) = (fillband!(H.data, x, -1, size(H,2)-1); H)
 
 +(A::UpperHessenberg, B::UpperHessenberg) = UpperHessenberg(A.data+B.data)
 -(A::UpperHessenberg, B::UpperHessenberg) = UpperHessenberg(A.data-B.data)
-# (future: we could also have specialized routines for UpperHessenberg ± UpperTriangular)
 
-# shift Hessenberg by λI
-+(H::UpperHessenberg, J::UniformScaling) = UpperHessenberg(H.data + J)
--(J::UniformScaling, H::UpperHessenberg) = UpperHessenberg(J - H.data)
+for T = (:UniformScaling, :Diagonal, :Bidiagonal, :Tridiagonal, :SymTridiagonal,
+         :UpperTriangular, :UnitUpperTriangular)
+    for op = (:+, :-)
+        @eval begin
+            $op(H::UpperHessenberg, x::$T) = UpperHessenberg($op(H.data, x))
+            $op(x::$T, H::UpperHessenberg) = UpperHessenberg($op(x, H.data))
+        end
+    end
+end
+
+for T = (:Number, :UniformScaling, :Diagonal)
+    @eval begin
+        *(H::UpperHessenberg, x::$T) = UpperHessenberg(H.data * x)
+        *(x::$T, H::UpperHessenberg) = UpperHessenberg(x * H.data)
+        /(H::UpperHessenberg, x::$T) = UpperHessenberg(H.data / x)
+        \(x::$T, H::UpperHessenberg) = UpperHessenberg(x \ H.data)
+    end
+end
+
+function *(H::UpperHessenberg, U::UpperOrUnitUpperTriangular)
+    T = _inner_type_promotion(H,U)
+    HH = similar(H.data, T, size(H))
+    copyto!(HH, H)
+    rmul!(HH, convert(AbstractArray{T}, U))
+    UpperHessenberg(HH)
+end
+function *(U::UpperOrUnitUpperTriangular, H::UpperHessenberg)
+    T = _inner_type_promotion(U,H)
+    HH = similar(H.data, T, size(H))
+    copyto!(HH, H)
+    lmul!(convert(AbstractArray{T}, U), HH)
+    UpperHessenberg(HH)
+end
+
+function /(H::UpperHessenberg, U::UpperTriangular)
+    T = typeof((zero(eltype(H))*zero(eltype(U)) + zero(eltype(H))*zero(eltype(U)))/one(eltype(H)))
```

Comment:
This seems dimensionally incorrect if these are unitful matrices.   Seems like `T = typeof(zero(eltype(H)) / oneunit(eltype(U)))` should suffice?

Reply:
Yes, this doesn’t work for unitful matrices. It was taken from
https://github.com/JuliaLang/julia/blob/027bd4a4d17979e5f78ca6831cac456dc41d1db8/stdlib/LinearAlgebra/src/triangular.jl#L1686
So it would need to be fixed there as well. I will make a separate PR for that.

Reply:
I don’t think this can be easily fixed in all cases: `HH` must be able to hold the original `H` and the result of `H/U`, so it will not work if `U` is unitful. However, with your proposed change it will at least work for unitful `H`, so I will implement that.

Reply:
I now changed the type computation to allow unitful matrices, but it still doesn’t work in all cases because some other methods (unrelated to `UpperHessenberg`) would need to be changed as well. For now, I would leave it at that. Support for dimensionful matrices in all `*`, `/`, `\` methods can be tackled in another PR.